### ISSUE TITLE
CoreDNS caching CNAME without an A record

### ISSUE THREAD

**Issue by nvergottini** [miekg/coredns#6958]
I recently switched my company to use CoreDNS as our primary DNS resolver with a new external DNS provider. The Corefile has multiple zone blocks that forward to our Windows domain controllers for internal zones and a root zone block that forwards to our external DNS provider. The last few nights we have run into an issue where a specific CNAME record for which our Windows DCs are authoritative fails to resolve to an A record. The CNAME record points to an external DNS record. Our Windows DCs are also set up to forward to the same external DNS provider, so they are able to resolve the CNAME to an A record and return an answer containing both the CNAME and the A record. Normally this works just fine, but for several nights in a row, it appears that the Windows DCs have been unable to resolve the A record for a short time (probably due to an upstream issue), so they have returned only the CNAME record to CoreDNS. CoreDNS itself does not attempt to resolve the CNAME record to the A record and returns only the CNAME to the client. It appears also that CoreDNS caches this non-answer, and keeps returning it to the clients until the TTL on the CNAME expires, which effectively makes the associated site unreachable.

I should add that I believe this is an issue with CoreDNS caching a bad answer because I was able to resolve by simply restarting CoreDNS.

Is this expected behavior? Should Windows DNS return a CNAME as NOERROR without the A record if it is unable to resolve the A record? Should CoreDNS cache what is effectively a useless answer? Is there a way for CoreDNS to attempt to resolve the CNAME itself if the upstream forwarder returns a CNAME without an A record?

**Comment by akhilsingh-git** [comment]
Yes, this is expected behavior given standard DNS semantics. Windows DNS can return a CNAME alone if it cannot resolve the final target at that moment, and CoreDNS will cache what it receives. Out-of-the-box, CoreDNS’s forward plugin will not “chase” the CNAME itself. You may need to adjust your configuration, use a recursive resolver, or ensure that your Windows DNS server does not return incomplete CNAME answers to avoid this issue.

**Comment by nitin-nizhawan** [comment]
This is fundamentally a **caching-correctness** bug, not merely "the `forward` plugin doesn't chase CNAMEs."

### Symptom

When an upstream returns `NOERROR` with a CNAME (or CNAME chain) but **no terminating A/AAAA record and no SOA** in the authority section (e.g. an incomplete recursion result from a forwarder such as a Windows DC), CoreDNS stores it in the **positive** cache keyed on `<qname, qtype>` for the min CNAME TTL, and then replays that incomplete answer to every client until it expires. Restarting CoreDNS "fixes" it only because that flushes the cache.

### Root cause (code)

`plugin/pkg/response/typify.go` classifies *any* `NOERROR` with a non-empty answer section as a positive reply:

```go
if len(m.Answer) > 0 && m.Rcode == dns.RcodeSuccess {
    return NoError, opt
}
```

This never checks whether the answer actually contains a record of the queried type or whether the CNAME chain terminates. So a CNAME-only (incomplete) answer is treated as a full positive answer. `plugin/cache/cache.go` `key()` then guards the no-SOA case **only** for `NameError`:

```go
if t == response.NameError && !hasSOA(m) {
    return false, 0
}
```

There is no equivalent guard for this `NOERROR` / no-answer-of-qtype / no-SOA case, so it gets cached in `pcache`.

### RFC basis

Per **RFC 2308 §2.2**, this is semantically a **NODATA** response (NOERROR, no relevant answer, empty/NS-less authority — §2.2 explicitly notes NODATA responses may carry CNAMEs, concluding NODATA for the last CNAME/QNAME). **RFC 2308 §5** states:

> Negative responses without SOA records SHOULD NOT be cached as there is no way to prevent the negative responses looping forever between a pair of servers even with a short TTL.

CoreDNS already enforces exactly this for NXDOMAIN, but sidesteps it here via the positive-classification loophole.

### Prior art

This was analyzed before in #4987 and #5077. In #5077 @chrisohaver agreed:

> Should we be caching at all in this scenario? Per RFC 2308 ... Negative responses without SOA records SHOULD NOT be cached
> ...
> Seems it would be better to not cache at all than to cache for less time, if that's what the RFC prescribes.

Both PRs went stale without a merge, so the behavior persists (this issue is the 2024 recurrence).

### Proposed fix

Aligned with the earlier maintainer consensus: in `key()`, do not cache a `NoError` response whose answer section is a non-empty CNAME chain that never terminates in the requested type and that lacks an SOA — i.e. extend the existing `NameError && !hasSOA` guard to cover this SOA-less pseudo-NODATA case. CoreDNS then re-queries upstream on the next request instead of serving a sticky incomplete answer.

I'll open a PR implementing this shortly and link it here.

**Comment by nitin-nizhawan** [comment]
Opened #8232 implementing the proposed fix: the cache plugin now skips caching a NOERROR response whose answer section is a non-terminating CNAME chain lacking an SOA (mirroring the existing NXDOMAIN no-SOA guard), so the next query is resolved upstream again instead of replaying the incomplete answer. Includes unit tests for the dangling-CNAME (not cached) case plus terminating-A and direct-CNAME positive controls.

**Comment by nitin-nizhawan** [comment]
checking in

**Comment by nitin-nizhawan** [comment]
checking in
